model/src/main/java/com/example/opencommerce/domain/pricing/ItemPrice.java

```diff
@@ -0,0 +1,185 @@
+    private final ItemId id;
+    private final List<HistoricalPrice> history = new ArrayList<>();
+    private Value basePriceNett;
+    private Value discount;
+    private Vat vat;
+    private boolean initiated;
+
+    private ItemPrice(ItemId id) {
+        this.id = requireNonNull(id);
+        this.basePriceNett = Value.ZERO;
+        this.vat = Vat.ZERO;
+        this.discount = Value.ZERO;
+        this.initiated = false;
+    }
+
+    public static ItemPrice create(ItemId id) {
+        return new ItemPrice(id);
+    }
+
+    public void when(List<PriceEvent> history) {
+        history.forEach(this::apply);
+    }
+
+    public OperationResult<NewPriceInitiatedEvent> initiate(Value priceNett, Vat vat) {
+        return executePriceInitiate(priceNett, vat);
+    }
+
+    public OperationResult<BasePriceChangedEvent> changeBasePrice(Value newPrice, Instant executionDate) {
+        return executePriceChange(newPrice, executionDate);
+    }
+
+    public OperationResult<DiscountAppliedEvent> addNewDiscount(Value discount, Instant executionDate) {
+        return executeDiscount(discount, executionDate);
+    }
+
+    public OperationResult<DiscountRemovedEvent> removeCurrentDiscount(Instant executionDate) {
+        return executeDiscountRemoval(executionDate);
+    }
+
+    private Value lowestPriceWithinPeriod(Period timePeriod) {
+        Instant startDate = Instant.now()
+                .minus(timePeriod);
+        return this.history.stream()
+                .filter(price -> price.getValidTo()
+                        .isAfter(startDate))
+                .min(Comparator.comparing(p -> p.getPriceNett()
+                        .asDecimal()))
+                .map(HistoricalPrice::getPriceNett)
+                .orElse(this.basePriceNett);
+    }
+
+    public PriceSnapshot getSnapshot() {
+        if (discounted()) {
+            return PriceSnapshot.discountedPrice(this.id, this.basePriceNett, this.discount, lowestPriceWithinPeriod(Period.ofDays(30)), this.vat);
+        } else {
+            return PriceSnapshot.regularPrice(this.id, this.basePriceNett, this.vat);
+        }
+    }
+
+    private void apply(PriceEvent e) {
+
+        OperationResult<?> result;
+        if (e instanceof NewPriceInitiatedEvent event) {
+            result = executePriceInitiate(Value.of(event.getBasePrice()), Vat.of(event.getVat()));
+        } else if (e instanceof BasePriceChangedEvent event) {
+            result = executePriceChange(Value.of(event.getNewPrice()), event.getValidFrom());
+        } else if (e instanceof DiscountAppliedEvent event) {
+            result = executeDiscount(Value.of(event.getDiscount()), event.getValidFrom());
+        } else if (e instanceof DiscountRemovedEvent event) {
+            result = executeDiscountRemoval(event.getValidFrom());
+        } else {
+            throw new IllegalStateException("Unknown price event type: " + e.getClass()
+                    .getSimpleName());
+        }
```

Comment:
Java 21 introduced what other languages already have, the pattern matching in switch statement:
https://openjdk.org/jeps/441
So this code could be similar to:
```
return switch (obj) {
        case Integer i -> String.format("int %d", i);
        case Long l    -> String.format("long %d", l);
        case Double d  -> String.format("double %f", d);
        case String s  -> String.format("String %s", s);
        default        -> obj.toString();
    };
```
the:
```
switch (e) {
case NewPriceInitiatedEvent event -> ...
```